resolver/src/lookup_state.rs

```diff
@@ -239,61 +240,54 @@ impl<C: DnsHandle<Error = ResolveError> + 'static> QueryFuture<C> {
 
     fn handle_noerror(&mut self, mut message: Message) -> Poll<Records, ResolveError> {
         // seek out CNAMES
-        // TODO: figure out how to get rid of this clone
-        let mut cname_ttl = 0;
-        let mut was_cname = false;
-        let mut search_name: Name = self.query.name().clone();
-        while let Some(cname) = message.answers().iter().find(|r| {
-            r.rr_type() == RecordType::CNAME && r.name() == &search_name
-        })
-        {
-            was_cname = true;
-            cname_ttl = cname.ttl();
-            if let &RData::CNAME(ref name) = cname.rdata() {
-                if search_name == *name {
-                    break; // already searched for this name
-                } else {
-                    search_name = name.clone();
-                }
-            } else {
-                // now that is very odd...
-                warn!("Expected RData::CNAME in response record {:?}", cname);
-                break;
+        let (search_name, cname_ttl, was_cname) = {
+            let (search_name, cname_ttl, was_cname) = message.answers().iter()
+                .fold((Cow::Borrowed(self.query.name()), 0, false),
+                      |(search_name, cname_ttl, was_cname), r| {
+                          if let &RData::CNAME(ref cname) = r.rdata() {
+                              debug_assert_eq!(r.rr_type(), RecordType::CNAME);
+                              if search_name.as_ref() == r.name() {
+                                  return (Cow::Owned(cname.clone()), r.ttl(), true);
+                              }
```

Comment:
Shouldn't we save each CNAME entry in this chain into the cache here? Otherwise we won't cache the CNAME, right? Imagine the case where the CNAME is very long lived and where the A records are very short lived. Then we'd requery the CNAME record unnecessarily many times, right?

Reply:
Perhaps, we should. The reason that's not happening now, is because of where I trigger the actual cache insertion, which happens post processing a lot of this. If we thread the cache in here, we'll need to add an additional state for the lookup state machine for inserting all CNAMEs after this call. CNAMEs that end up in requiring an additional query, would end up getting cached. See the result of the QueryState::CnameChain state of the QueryState machine on what was line 592.

The cache uses the lookup Query as the key, not (Names,RecordType) -> RecordSets. In effect it's acting as an L3 cache, we can discuss making it operate as a lower level record cache, but that would mean adding a lot of lookup logic to the cache for queries, I think. You can see Authority in Server as an example of what it would probably end up looking like. The cache being based on Query means that we won't rerun this CNAME logic, but we may end up duplicating some RecrodSets if they Query was formed differently, but returns the same RecordSet in the end as it is today.